### PR TITLE
🔥 Remove `_base_class_defined` hack in favor of an empty bases check

### PR DESCRIPTION
**Pull request by lig**
## Change Summary

Remove `_base_class_defined` hack and check for empty `bases` as an indicator for detecting `BaseModel` in `ModelMetaclass`. These classes are no longer tied to their places in the module.

## Related issue number

None 

## Checklist

* [X] Unit tests for the changes exist
* [x] Tests pass on CI and coverage remains at 100%
* [ ] Documentation reflects the changes where applicable
* [ ] `changes/<pull request or issue id>-<github username>.md` file added describing change
  (see [changes/README.md](https://github.com/samuelcolvin/pydantic/blob/master/changes/README.md) for details)
* [x] My PR is ready to review, **please add a comment including the phrase "please review" to assign reviewers**

Selected Reviewer: @adriangb

### CODE REVIEW

**Comment by lig** [comment]
please review

File: pydantic/main.py
Comment (dmontagu) [inline pydantic/main.py:1320]:
```suggestion
```
It seems to me this is unnecessary since the only place where this is called already checks for this conditional and bails out. (And all tests pass for me locally without these two lines.)

**Comment by dmontagu** [comment]
I'll just note this may be in tension with some of the goals of https://github.com/pydantic/pydantic/pull/5740 — simplifying the MRO etc.

I like getting rid of the hacky check for subclass of BaseModel, but I don't have a strong opinion about what's best here (or whether there might be other unintended consequences, especially for performance). Happy to defer to @adriangb or @samuelcolvin

File: pydantic/main.py
Comment (lig) [reply pydantic/main.py:1320]: Yes. I was just a bit paranoid. And there was an old check there. So, I'm not oppose to removing this but would like to have a some kind of a check here. Maybe an `assert` like `assert(bases != (_BaseModelBase,))`?

**Comment by lig** [comment]
> I'll just note this may be in tension with some of the goals of #5740 — simplifying the MRO etc.

I think, it would be even better. I'm happy to remove `_BaseModelBase` if `ModelBase` doesn't have any other bases.

File: pydantic/main.py
Comment (lig) [reply pydantic/main.py:1320]: BTW, this could be safely removed if `BaseModel` has no bases.

File: pydantic/main.py
Comment (samuelcolvin) [inline pydantic/main.py:108]: can we change this to `if bases != ...` and then revert most of these changes here?

File: pydantic/main.py
Comment (samuelcolvin) [inline pydantic/main.py:214]: I assume this is here to be used by `RootModel`? Otherwise I'd rather not have it.

**Review by samuelcolvin — COMMENTED** [review]
I'm okay with this in principle, but better if we can reduce the size of the change.

**Review by samuelcolvin — APPROVED** [review]
LGTM. @adriangb and @dmontagu are you happy with this?

File: pydantic/main.py
Comment (dmontagu) [inline pydantic/main.py:1336]:
```suggestion
        if base == BaseModel:
            continue
```
(then dedent the remaining bits)

File: pydantic/main.py
Comment (dmontagu) [inline pydantic/main.py:108]: It seems to me the comment above this line should be updated a bit. At least to say something like "`BaseModel` itself won't have any bases, but any subclass of it will"

**Review by dmontagu — APPROVED** [review]
I'm good with the change, I just think the big comment that seems to have been moved verbatim probably deserves an update before merging.

File: pydantic/main.py
Comment (lig) [reply pydantic/main.py:107]: Updated the comment

File: pydantic/main.py
Comment (dmontagu) [inline pydantic/main.py:105]:
```suggestion
        # that `BaseModel` itself won't have any bases, but any subclass of it will, to determine whether the `__new__`
```
I think this reads a bit better since the next line checks that it's _not_ the BaseModel class, as phrased it reads to me like it's confirming we are doing BaseModel, which confused me for a sec.

**Pull request merged**
